### PR TITLE
[huggingface] Add `torch.distributed` support for Trainium and `torchrun`

### PR DESCRIPTION
**Pull request by philschmid**
*Description of changes:*

This PR adds the `torch_distrubted` support to the `HuggingFace` estimator. The code is mostly copied from the already existing implementation for it in the `PyTorch` estimator. 

*Testing done:*

local tests and added a new test `test_huggingface_torch_distributed.py`

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by claytonparnell — APPROVED** [review]
/bot run all

**Comment by claytonparnell** [comment]
Hi @philschmid - I see there are some formatting errors, could you please run `tox -e black-format` and `tox -e flake8` to resolve?

**Comment by philschmid** [comment]
> Hi @philschmid - I see there are some formatting errors, could you please run `tox -e black-format` and `tox -e flake8` to resolve?

@claytonparnell  Fixed! Sorry for that.

**Review by claytonparnell — COMMENTED** [review]
/bot run all

File: tests/integ/test_huggingface_torch_distributed.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+from __future__ import absolute_import
+
+import os
+import pytest
+from sagemaker.huggingface import HuggingFace
+from tests.integ import timeout
+from tests.integ import DATA_DIR, TRAINING_DEFAULT_TIMEOUT_MINUTES
+
+
+@pytest.mark.release
```
Comment (claytonparnell) [inline tests/integ/test_huggingface_torch_distributed.py:22]: Could we remove the mark as release? This makes test run in all regions, and I worry that instance type may not be available in all regions. Test will still run in us-west-2.

Comment (claytonparnell) [reply tests/integ/test_huggingface_torch_distributed.py:22]: Sorry I didn't catch this earlier, once this is removed I can make sure it's merged this week

**Review by claytonparnell — COMMENTED** [review]
/bot run all

**Comment by philschmid** [comment]
@claytonparnell fixed the flake8 error. can you run the tests again

**Review by claytonparnell — COMMENTED** [review]
/bot run all

**Comment by claytonparnell** [comment]
@philschmid can you look at [this link](https://sxsf0yqucb.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=a450d5fb-63e2-4014-8a4a-bfe94c10e6b1%2Fbuild.log) and look into why the newly added test is failing?

**Comment by philschmid** [comment]
> @philschmid can you look at [this link](https://sxsf0yqucb.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=a450d5fb-63e2-4014-8a4a-bfe94c10e6b1%2Fbuild.log) and look into why the newly added test is failing?

Fixed the `test_huggingface_torch_distributed_g5_glue` test and update the branch to be in sync with master. 

The logs had some unrelated failures 
```bash
FAILED tests/integ/test_marketplace.py::test_marketplace_attach - botocore.ex...
FAILED tests/integ/test_marketplace.py::test_marketplace_transform_job - boto...
FAILED tests/integ/test_marketplace.py::test_marketplace_tuning_job - botocor...
```

i hope those are fixed by syncing the main branch. 

Can you rerun the tests?

File: tests/integ/test_huggingface_torch_distributed.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    with timeout.timeout(minutes=TRAINING_DEFAULT_TIMEOUT_MINUTES):
+        data_path = os.path.join(DATA_DIR, "huggingface")
+        estimator = HuggingFace(
+            py_version=huggingface_pytorch_latest_training_py_version,
+            entry_point=os.path.join(data_path, "run_glue.py"),
+            role="SageMakerRole",
+            transformers_version=huggingface_training_latest_version,
+            pytorch_version=huggingface_training_pytorch_latest_version,
+            instance_count=1,
+            instance_type="g5.12xlarge",
```
Comment (claytonparnell) [inline tests/integ/test_huggingface_torch_distributed.py:35]: @philschmid Can you change to `ml.g5.12xlarge` ? Currently test `test_huggingface_torch_distributed_g5_glue` is failing due to regex not recognizing instance type

Comment (philschmid) [reply tests/integ/test_huggingface_torch_distributed.py]: fixed. Sorry my bad

Comment (claytonparnell) [reply tests/integ/test_huggingface_torch_distributed.py]: No worries! Sorry didn't catch earlier

**Review by claytonparnell — COMMENTED** [review]
/bot run all

**Pull request merged**